File: apps/mobile-flutter/test/unit/session_and_settings_test.dart
import 'package:careconnect_mobile/core/utils/clock.dart';
import 'package:careconnect_mobile/data/local_store.dart';
import 'package:careconnect_mobile/models/notification_settings.dart';
import 'package:careconnect_mobile/models/session.dart';
import 'package:careconnect_mobile/models/user_role.dart';
import 'package:careconnect_mobile/state/clock_provider.dart';
import 'package:careconnect_mobile/state/notification_settings_provider.dart';
import 'package:careconnect_mobile/state/session_provider.dart';
import 'package:careconnect_mobile/state/settings_provider.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../helpers/test_app.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  group('session', () {
    test('starts signed out and defaults the role to care recipient', () async {
      final container = await createContainer();
      expect(container.read(sessionProvider), isNull);
      expect(container.read(currentRoleProvider), UserRole.careRecipient);
    });

    test('signIn persists and signOut clears', () async {
      final container = await createContainer();
      await container
          .read(sessionProvider.notifier)
          .signIn(
            role: UserRole.caregiver,
            method: SignInMethod.email,
            email: '  renee@example.test ',
          );
      final session = container.read(sessionProvider)!;
      expect(session.role, UserRole.caregiver);
      expect(session.email, 'renee@example.test');
      expect(session.signedInAt, kTestNow);
      expect(container.read(currentRoleProvider), UserRole.caregiver);
      final prefs = await SharedPreferences.getInstance();
      expect(prefs.containsKey(StoreKeys.session), isTrue);

      await container.read(sessionProvider.notifier).signOut();
      expect(container.read(sessionProvider), isNull);
      expect(prefs.containsKey(StoreKeys.session), isFalse);
    });

    test('blank emails are stored as null', () async {
      final container = await createContainer();
      await container
          .read(sessionProvider.notifier)
          .signIn(
            role: UserRole.careRecipient,
            method: SignInMethod.passkey,
            email: '   ',
          );
      expect(container.read(sessionProvider)!.email, isNull);
    });

    test('restores a persisted session', () async {
      final container = await createContainer(
        prefs: sessionPrefs(UserRole.caregiver),
      );
      expect(container.read(sessionProvider)!.role, UserRole.caregiver);
    });
  });

  group('app settings', () {
    test('defaults and persistence', () async {
      final container = await createContainer();
      expect(container.read(appSettingsProvider), const AppSettings());
      expect(container.read(themeModeProvider), ThemeMode.system);

      final notifier = container.read(appSettingsProvider.notifier);
      await notifier.setThemeMode(ThemeMode.dark);
      await notifier.setShareWithCaregiver(enabled: false);
      expect(container.read(themeModeProvider), ThemeMode.dark);
      expect(container.read(appSettingsProvider).shareWithCaregiver, isFalse);

      final prefs = await SharedPreferences.getInstance();
      final reloaded = await createContainer(
        prefs: {StoreKeys.settings: prefs.getString(StoreKeys.settings)!},
      );
      expect(reloaded.read(appSettingsProvider).themeMode, ThemeMode.dark);
      expect(reloaded.read(appSettingsProvider).shareWithCaregiver, isFalse);
    });

    test('JSON ignores unknown theme modes', () {
      expect(
        AppSettings.fromJson(const {'themeMode': 'sepia'}).themeMode,
        ThemeMode.system,
      );
      const a = AppSettings(demoClock: true);
      expect(a.hashCode, AppSettings.fromJson(a.toJson()).hashCode);
    });
  });

  group('clock', () {
    test('follows the injected clock', () async {
      final clock = MutableClock(kTestNow);
      final container = await createContainer(clock: clock);
      expect(container.read(currentTimeProvider), kTestNow);
      clock.current = kTestNow.add(const Duration(minutes: 5));
      container.read(currentTimeProvider.notifier).refresh();
      expect(
        container.read(currentTimeProvider),
        kTestNow.add(const Duration(minutes: 5)),
      );
    });

    test('demo clock freezes time at the design instant', () async {
      final container = await createContainer(clock: FixedClock.at(2030, 1, 1));
      await container
          .read(appSettingsProvider.notifier)
          .setDemoClock(enabled: true);
      expect(container.read(currentTimeProvider), kDemoInstant);
      container.read(currentTimeProvider.notifier).refresh();
      expect(container.read(currentTimeProvider), kDemoInstant);
      await container
          .read(appSettingsProvider.notifier)
          .setDemoClock(enabled: false);
      expect(container.read(currentTimeProvider), DateTime(2030));
    });

    test('SystemClock reports wall time', () {
      final before = DateTime.now();
      final now = const SystemClock().now();
      expect(now.isBefore(before), isFalse);
    });
  });

  group('notification settings', () {
    test('defaults, updates and persistence', () async {
      final container = await createContainer();
      expect(
        container.read(notificationSettingsProvider),
        const NotificationSettings(),
      );
      final notifier = container.read(notificationSettingsProvider.notifier);
      await notifier.setDailyDigest(enabled: false);
      await notifier.setLeadTime(ReminderLeadTime.fifteenMinutes);
      final prefs = await SharedPreferences.getInstance();
      final reloaded = await createContainer(
        prefs: {
          StoreKeys.notifications: prefs.getString(StoreKeys.notifications)!,
        },
      );
      final settings = reloaded.read(notificationSettingsProvider);
      expect(settings.dailyDigest, isFalse);
      expect(settings.leadTime, ReminderLeadTime.fifteenMinutes);
    });
  });
}

File: apps/react-mobile/src/state/sessionAndSettings.test.ts
import AsyncStorage from '@react-native-async-storage/async-storage';

import { StoreKeys } from '../data/localStore';
import { defaultAppSettings, defaultNotificationSettings } from '../models/serialization';
import { FixedClock } from '../core/utils/clock';
import { useClockStore } from './clockStore';
import { useNotificationSettingsStore } from './notificationSettingsStore';
import { useSessionStore } from './sessionStore';
import { useSettingsStore } from './settingsStore';

beforeEach(async () => {
  await AsyncStorage.clear();
  useSessionStore.setState({ session: null, hydrated: true });
  useSettingsStore.setState({ ...defaultAppSettings(), hydrated: true });
  useNotificationSettingsStore.setState({ ...defaultNotificationSettings(), hydrated: true });
  useClockStore.getState().setClock(new FixedClock(new Date(2026, 7, 25, 9, 0)));
});

describe('sessionStore', () => {
  it('signs in with a trimmed email and persists the session', async () => {
    await useSessionStore.getState().signIn({
      role: 'careRecipient',
      method: 'email',
      email: '  muhammad@example.test  ',
    });
    const session = useSessionStore.getState().session!;
    expect(session.role).toBe('careRecipient');
    expect(session.email).toBe('muhammad@example.test');
    expect(session.signedInAt).toEqual(new Date(2026, 7, 25, 9, 0));

    const stored = await AsyncStorage.getItem(StoreKeys.session);
    expect(JSON.parse(stored!).role).toBe('careRecipient');
  });

  it('stores a null email for a passkey sign-in', async () => {
    await useSessionStore.getState().signIn({ role: 'caregiver', method: 'passkey' });
    expect(useSessionStore.getState().session?.email).toBeNull();
    expect(useSessionStore.getState().session?.method).toBe('passkey');
  });

  it('signs out and clears storage', async () => {
    await useSessionStore.getState().signIn({ role: 'caregiver', method: 'passkey' });
    await useSessionStore.getState().signOut();
    expect(useSessionStore.getState().session).toBeNull();
    expect(await AsyncStorage.getItem(StoreKeys.session)).toBeNull();
  });
});

describe('settingsStore', () => {
  it('updates and persists theme mode', async () => {
    await useSettingsStore.getState().setThemeMode('dark');
    expect(useSettingsStore.getState().themeMode).toBe('dark');
    const stored = await AsyncStorage.getItem(StoreKeys.settings);
    expect(JSON.parse(stored!).themeMode).toBe('dark');
  });

  it('toggling the demo clock refreshes the clock store', async () => {
    expect(useClockStore.getState().now).toEqual(new Date(2026, 7, 25, 9, 0));
    await useSettingsStore.getState().setDemoClock(true);
    expect(useClockStore.getState().now).toEqual(new Date(2026, 7, 25, 14, 14));
    await useSettingsStore.getState().setDemoClock(false);
    expect(useClockStore.getState().now).toEqual(new Date(2026, 7, 25, 9, 0));
  });

  it('updates caregiver sharing', async () => {
    await useSettingsStore.getState().setShareWithCaregiver(false);
    expect(useSettingsStore.getState().shareWithCaregiver).toBe(false);
  });
});

describe('notificationSettingsStore', () => {
  it('updates and persists the daily digest and lead time', async () => {
    await useNotificationSettingsStore.getState().setDailyDigest(false);
    await useNotificationSettingsStore.getState().setLeadTime('oneDay');
    expect(useNotificationSettingsStore.getState().dailyDigest).toBe(false);
    expect(useNotificationSettingsStore.getState().leadTime).toBe('oneDay');
    const stored = await AsyncStorage.getItem(StoreKeys.notifications);
    expect(JSON.parse(stored!)).toEqual({ dailyDigest: false, leadTime: 'oneDay' });
  });
});
